Feature add track

w96j0 opened the pull request.

w96j0 committed:
Remove unused properties
Add application.yml for db configuration to test

w96j0 committed:
Add method to addTrack

w96j0 committed:
Merge branch 'main' into feature-add-track

# Conflicts:
#	Backend/src/test/resources/application.yml

w96j0 committed:
Add integration test for adding track

w96j0 committed:
Add validation to validate the value of name not blank
Add test to test the validation

w96j0 committed:
Add fields to add new track in the frontend

w96j0 committed:
Add method to avoid duplicate song

w96j0 committed:
Add test to check duplicate song
Add GlobalExceptionHandler to catch the IllegalArgumentException

w96j0 committed:
Remove unused TrackServiceTest file

matthias-hiemer commented on Backend/src/main/java/com/valantic/myplaylist/Track.java:
```diff
@@ -22,6 +23,7 @@ public class Track {
     private Integer id;
+    @NotBlank
     private String name;
     private String artist;
```
Ich glaube es wäre besser wenn wir auch artist validieren würden.
Bin mir aber nicht zu 100% sicher. Was spricht dafür und was dagegen?

```suggestion
    @NotBlank
    private String artist;
```

matthias-hiemer commented on Backend/src/main/java/com/valantic/myplaylist/TrackController.java:
```diff
@@ -19,4 +18,9 @@ public TrackController(TrackService trackService) {
     }
+
+    @PostMapping
+    public void addTrack(@Valid @RequestBody Track newTrack) {
```
Wenn du willst:
Gib hier den neu hinzugefügten Track zurück

w96j0 replied:
create a method in TrackService to get the specific track by id
to get the added track (?) back from db after adding new track
TODO:  test the return value

matthias-hiemer commented on Backend/src/main/java/com/valantic/myplaylist/TrackService.java:
```diff
@@ -15,4 +15,12 @@ public TrackService(TrackRepository trackRepository) {
+
+    public void addTrack(Track newTrack) {
+        if(trackRepository.existsByNameAndArtist(newTrack.getName(), newTrack.getArtist())) {
+            throw new IllegalArgumentException("The Song is already existing!");
```
Gib hier Name und Artist zusätzlich aus.
Das vereinfacht später auch die Fehlersuche im Log.

Regel ist: Beim Loggen immer möglichst viel (sinnvollen) Kontext mitliefern/mitloggen.
Das vereinfacht die Wartung und Fehlersuche.

matthias-hiemer commented on Backend/src/test/java/com/valantic/myplaylist/TrackControllerTest.java:
```diff
@@ -47,4 +49,79 @@ void getTracks() throws Exception {
+
+    @Test
+    void test_addTrack_Avoid_Duplicate_Song() throws Exception {
+        Track defalutTrack = new Track(
```
Kleiner Typo :)

matthias-hiemer commented on Backend/src/test/java/com/valantic/myplaylist/TrackControllerTest.java:
```diff
@@ -47,4 +49,79 @@ void getTracks() throws Exception {
+        String testTrackJson = """ 
+                {
+                "id":1,
+                "name":,
```
```suggestion
                "name": "",
```

matthias-hiemer commented on Backend/src/test/java/com/valantic/myplaylist/TrackControllerTest.java:
```diff
@@ -47,4 +49,79 @@ void getTracks() throws Exception {
+                .andDo(print())
+                .andExpect(status().isOk());
+
+        assertTrue(trackRepository.existsById(1));
```
Hier kannst du stattdessen auch prüfen:
existsBy: Name + Artist

Damit bist du nicht von der Id abhängig

matthias-hiemer commented on Frontend/src/components/Playlist.vue:
```diff
@@ -9,9 +9,11 @@
-      <input v-model="newTrack.name" placeholder="Name" />
-
-      <input v-model="newTrack.artist" placeholder="Artist" />
+      <input v-model="newTrack.name" placeholder="Name"/>
```
Sieht gut aus.
Teste es noch im Frontend :)

matthias-hiemer reviewed (changes requested):
siehe Code Kommentare.
Sieht aber gut aus :)

w96j0 committed:
Add validation for artist
Add test for the validation
Change the addTrack method to return added track as value

w96j0 committed:
Add name and artist of song to the exception message when a song is already existing
Correct the typo

w96j0 committed:
Add file TrackServiceTest to test the getTrack

w96j0 committed:
Add to check the return value

matthias-hiemer commented on Backend/src/test/java/com/valantic/myplaylist/TrackServiceTest.java:
```diff
@@ -0,0 +1,46 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+@SpringBootTest
```
Diese Zeile brauchen wir nur in Controller Tests. (Damit sagen wir Spring: fahre komplett Spring hoch, erstelle alle Dependencies, usw.). Das wollen wir an diese Stelle nicht:
Grund 1: Performance. das dauert sehr lange
Grund 2: Dadurch könnten wir mit @Autowired Dependencies injecten. Aber! Genau wollen wir nicht in Unit Tests.

Hier bitte löschen

w96j0 committed:
Remove the springboottest dependency
